### DISCUSSION TITLE
CLI Text Printing

### DISCUSSION THREAD

**Discussion by NaitLee** [NaitLee/Cat-Printer#11] [Ideas]
Command-line text printing defaults to use ImageMagick. Below is the way to do with PF2 font — lightweight, energy efficient, and streamed, getting much flexibility.

To try it out, do these:

0. Gather newest release.
1. Get a PF2 font file, put to repo folder as `font.pf2`
  For GNU/Linux you have a big change to find a `unifont.pf2` at somewhere in directory `/boot/grub`.
  Or take this: [unifont.pf2.zip](https://github.com/NaitLee/Cat-Printer/files/8561846/unifont.pf2.zip)
  Otherwise, generate one with command `grub-mkfont` along with your favorite TTF/OTF font. It's very customizable.
  Unfortunately, other OS like Windows have no official pre-distributed `grub-mkfont`. I’ve failed to compile one, either.
2. Enable bluetooth and powerup printer, invoke `python3 printer.py -t 1,font,pf2 text_file.txt`

Well, that’s all. In this case no complex libs required. If you'd like to know what's happening, check file `printer_lib/pf2.py`, `printer_lib/text_print.py` and the updated `printer.py`

Try the “typewriter” hack! Give `-` as filename and type directly in the terminal, line by line.

Cheers~

**Comment by frankenstein91** [comment]
Hi
I testet with `echo "Das ist nur ein Test mit mehr Zeichen als der erste Test und so"|python3 printer.py -t -` and the Output is not that nice.
The is a lot of space between the lines. Also some pixels of the font are missing

**Reply by frankenstein91** [reply]
missing pixels is not with every font... but Gamecuben has that problem

**Reply by NaitLee** [reply]
Usually fancy fonts have fancy use of font features. In Gamecuben it's (maybe) character "descend" been set too low, it looks well in normal editors (that allow rendering outside the box), but current version of our engine is just a bit silly...

In the short future I'll test & do more.

Edit: remove obsolete patch

**Reply by claudiorrrr** [reply]
It does work to me using an Spanish language text (I'm working for also translating the repo to Spanish, more to come)

But I couldn't find the way to manually change the font size. Is there a way?

**Reply by NaitLee** [reply]
Ah, well, PF2 font size is fixed.
I may do font zooming later. That's relatively easy, but maybe only 2x, 3x etc.
I will also change cmd parameter behavior, make all text-printing params defined all-together, for convenience.
And also frontend text printing in the future.
But as said you can also generate your own, `grub-mkfont` can customize size.
I failed to compile grub for windows. ~~Let me try if Mac is possible.~~

Thank you for translation 🎉

**Reply by claudiorrrr** [reply]
No news for a 2x zoom on the font? that would be wicked

**Reply by NaitLee** [reply]
> No news for a 2x zoom on the font? that would be wicked

See newest 0.4, for text printing it enables the use of another different technique and you can do much more with that :)
I'm gonna publish a fix in few minutes. ~~Please wait for a moment...~~ Done!
Also you might should reply to this discussion (at bottom of page) rather than to a comment :)

PS. I've been freaked out because of other matter these days and did less things. If you really mean PF2 zooming then I'd do it in following days.

**Comment by NaitLee** [comment]
With [ImageMagick](https://imagemagick.org/) we are able to convert from png/jpg to pbm.
But we can also [make text image](https://legacy.imagemagick.org/Usage/text/) with it -- then print.
Though it's also possible to make pf2 font with it.
Let me do the "integration".

**Comment by NaitLee** [comment]
See newest 0.4.0. It uses ImageMagick to print text, and it's able to define font size & font family, freely.
PF2 is still there for good reasons:
- Truly lightweight
- Could stream the text IO, by line
- Do things more quickly
- By flipping the direction you can print streamed text (lines) in reverse order

**Comment by NaitLee** [comment]
**Finally** PF2 text zooming (glyph scaling) is there, in 0.5.2.

It's actually little work. I just compromised doing something else these days...
